network.go

```diff
@@ -0,0 +1,505 @@
+	var cleanupFuncs []func() error
+
+	// Get the network interface with CNI configuration or, if there is none,
+	// just return right away.
+	cniNetworkInterface := networkInterfaces.cniInterface()
+	if cniNetworkInterface == nil {
+		return nil, cleanupFuncs
+	}
+
+	cniNetworkInterface.CNIConfiguration.netNSPath = netNSPath
+	cniNetworkInterface.CNIConfiguration.setDefaults(machineID)
+
+	// Make sure the netns is setup. If the path doesn't yet exist, it will be
+	// initialized with a new empty netns.
+	err, netnsCleanupFuncs := cniNetworkInterface.CNIConfiguration.initializeNetNS()
+	cleanupFuncs = append(cleanupFuncs, netnsCleanupFuncs...)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize netns"), cleanupFuncs
+	}
+
+	cniResult, err, cniCleanupFuncs := cniNetworkInterface.CNIConfiguration.invokeCNI(ctx, logger)
+	cleanupFuncs = append(cleanupFuncs, cniCleanupFuncs...)
+	if err != nil {
+		return errors.Wrap(err, "failure when invoking CNI"), cleanupFuncs
+	}
+
+	// If static configuration is not already set for the network device, fill it out
+	// by parsing the CNI result object according to the specifications detailed in the
+	// vmconf package docs.
+	if cniNetworkInterface.StaticConfiguration == nil {
+		vmNetConf, err := vmconf.StaticNetworkConfFrom(*cniResult, cniNetworkInterface.CNIConfiguration.ContainerID)
+		if err != nil {
+			return errors.Wrap(err,
+				"failed to parse VM network configuration from CNI output, ensure CNI is configured with a plugin "+
+					"that supports automatic VM network configuration such as tc-redirect-tap",
+			), cleanupFuncs
+		}
+
+		cniNetworkInterface.StaticConfiguration = &StaticNetworkConfiguration{
+			HostDevName: vmNetConf.TapName,
+			MacAddress:  vmNetConf.VMMacAddr,
+		}
+
+		if vmNetConf.VMIPConfig != nil {
+			cniNetworkInterface.StaticConfiguration.IPConfiguration = &IPConfiguration{
+				IPAddr:      vmNetConf.VMIPConfig.Address,
+				Gateway:     vmNetConf.VMIPConfig.Gateway,
+				Nameservers: vmNetConf.VMNameservers,
+			}
+		}
+	}
+
+	return nil, cleanupFuncs
+}
+
+// return the network interface that has CNI configuration, or nil if there is no such interface
+func (networkInterfaces NetworkInterfaces) cniInterface() *NetworkInterface {
+	// Validation that there is at most one CNI interface is done as part of the
+	// NetworkConfigValidationHandler, can safely just use the first result
+	// here and assume it's the only one.
+	for i, iface := range networkInterfaces {
+		if iface.CNIConfiguration != nil {
+			return &networkInterfaces[i]
+		}
+	}
+
+	return nil
+}
+
+// return the network interface that has static IP configuration, or nil if there is no such interface
+func (networkInterfaces NetworkInterfaces) staticIPInterface() *NetworkInterface {
+	// Validation that there is at most one interface with StaticIPConfiguration
+	// is done as part of the NetworkConfigValidationHandler, can safely just use
+	// the first result here and assume it's the only one.
```

Comment:
And if there are more than one? Should we return an error or do some more stricter validation up top?

Reply:
The code as is returns an error in the case that there is an interface with IP configuration when multiple interfaces were provided (see `NetworkInterfaces.validate`, currently line 360 of this file). The comment here was just meant to explain that validation already occurs there so we can thus operate on simpler assumptions rather than re-validate here.

Reply:
Yea, but this doesn't check if there aren't multiple `StaticConfiguration`s, which is where the issues lies for me. I don't know I like assuming that only one was provided.

Reply:
We aren't assuming there is only one provided, we are just doing the validation elsewhere. It's just up to the caller (in this case, `Machine`) to validate `NetworkInterfaces` before trying to use it. I considered making a `NewNetworkInterfaces` func that does the creation of `NetworkInterfaces` and validation all in one go, but that seemed to go against the pattern already in the SDK to use validation handlers, which is why I went with what I have.